mac/mcompile/mcompile.cpp

```diff
@@ -0,0 +1,597 @@
+      key->ShiftFlags &= ~VIRTUALCHARKEY;
+    }
+
+    int len = u16len(key->dpContext);
+
+    PKMX_WCHAR context = new KMX_WCHAR[len + 4];
+    memcpy(context, key->dpContext, len * sizeof(KMX_WCHAR));
+    context[len] = UC_SENTINEL;
+    context[len + 1] = CODE_DEADKEY;
+    context[len + 2] = deadkey;
+    context[len + 3] = 0;
+    key->dpContext = context;
+    key->Key = vk;
+  }
+}
+
+/**
+ * @brief  translate a group
+ * @param  group     pointer to a keyboard group
+ * @param  a         deadkey to be remapped
+ * @param  vk        a keyvalue of the US keyboard
+ * @param  shift     shiftstate
+ * @param  character of the underlying keyboard
+ */
+void mac_KMX_TranslateDeadkeyGroup(LPKMX_GROUP group, KMX_WCHAR deadkey, KMX_WORD vk, KMX_DWORD shift, KMX_WORD ch) {
+  for (unsigned int i = 0; i < group->cxKeyArray; i++) {
+    mac_KMX_TranslateDeadkeyKey(&group->dpKeyArray[i], deadkey, vk, shift, ch);
+  }
+}
+
+/**
+ * @brief  translate a keyboard
+ * @param  kbd       pointer to the US keyboard
+ * @param  a         deadkey to be remapped
+ * @param  vk        a keyvalue of the US keyboard
+ * @param  shift     shiftstate
+ * @param  character of the underlying keyboard
+ */
+void mac_KMX_TranslateDeadkeyKeyboard(LPKMX_KEYBOARD kbd, KMX_WCHAR deadkey, KMX_WORD vk, KMX_DWORD shift, KMX_WORD ch) {
+  for (unsigned int i = 0; i < kbd->cxGroupArray; i++) {
+    if (kbd->dpGroupArray[i].fUsingKeys) {
+      mac_KMX_TranslateDeadkeyGroup(&kbd->dpGroupArray[i], deadkey, vk, shift, ch);
+    }
+  }
+}
+
+/**
+ * @brief  add a deadkey rule
+ * @param  kbd     pointer to the US keyboard
+ * @param  deadkey a deadkey to be added
+ * @param  vk      a keyvalue of the US keyboard
+ * @param  shift   shiftstate
+ */
+void mac_KMX_AddDeadkeyRule(LPKMX_KEYBOARD kbd, KMX_WCHAR deadkey, KMX_WORD vk, KMX_DWORD shift) {
+  // The weird LCTRL+RALT is Windows' way of mapping the AltGr key.
+  // We store that as just RALT, and use the option "Simulate RAlt with Ctrl+Alt"
+  // to provide an alternate..
+  if ((shift & (LCTRLFLAG | RALTFLAG)) == (LCTRLFLAG | RALTFLAG))  // I4549
+    shift &= ~LCTRLFLAG;
+  // If the first group is not a matching-keys group, then we need to add into
+  // each subgroup, otherwise just the match group
+  for (unsigned int i = 0; i < kbd->cxGroupArray; i++) {
+    if (kbd->dpGroupArray[i].fUsingKeys) {
+      LPKMX_KEY keys = new KMX_KEY[kbd->dpGroupArray[i].cxKeyArray + 1];
+      memcpy(keys + 1, kbd->dpGroupArray[i].dpKeyArray, kbd->dpGroupArray[i].cxKeyArray * sizeof(KMX_KEY));
+      keys[0].dpContext = new KMX_WCHAR[1];
+      keys[0].dpContext[0] = 0;
+      keys[0].dpOutput = new KMX_WCHAR[4];
+      keys[0].dpOutput[0] = UC_SENTINEL;
+      keys[0].dpOutput[1] = CODE_DEADKEY;
+      keys[0].dpOutput[2] = deadkey; // TODO: translate to unique index
```

Comment:
Is this TODO resolved?

Reply:
this is a comment from mcompile-windows - I did not do any changes